File: ggp-base/src/org/ggp/base/util/gdl/factory/GdlFactory.java
package org.ggp.base.util.gdl.factory

import java.util.ArrayList
import java.util.List

import org.ggp.base.util.gdl.factory.exceptions.GdlFormatException
import org.ggp.base.util.gdl.grammar.Gdl
import org.ggp.base.util.gdl.grammar.GdlConstant
import org.ggp.base.util.gdl.grammar.GdlDistinct
import org.ggp.base.util.gdl.grammar.GdlFunction
import org.ggp.base.util.gdl.grammar.GdlLiteral
import org.ggp.base.util.gdl.grammar.GdlNot
import org.ggp.base.util.gdl.grammar.GdlOr
import org.ggp.base.util.gdl.grammar.GdlPool
import org.ggp.base.util.gdl.grammar.GdlProposition
import org.ggp.base.util.gdl.grammar.GdlRelation
import org.ggp.base.util.gdl.grammar.GdlRule
import org.ggp.base.util.gdl.grammar.GdlSentence
import org.ggp.base.util.gdl.grammar.GdlTerm
import org.ggp.base.util.gdl.grammar.GdlVariable
import org.ggp.base.util.symbol.factory.SymbolFactory
import org.ggp.base.util.symbol.factory.exceptions.SymbolFormatException
import org.ggp.base.util.symbol.grammar.Symbol
import org.ggp.base.util.symbol.grammar.SymbolAtom
import org.ggp.base.util.symbol.grammar.SymbolList


class GdlFactory


    def static Gdl create(String string) throws GdlFormatException, SymbolFormatException
	
        return create(SymbolFactory.create(string))

    def static Gdl create(Symbol symbol) throws GdlFormatException
	
        try
		
            return createGdl(symbol)
        catch (Exception e)
		
            createGdl(symbol)
            throw new GdlFormatException(symbol)

    def GdlConstant createConstant(SymbolAtom atom)
	
        return GdlPool.getConstant(atom.getValue())

    def GdlDistinct createDistinct(SymbolList list)
	
        GdlTerm arg1 = createTerm(list.get(1))
        GdlTerm arg2 = createTerm(list.get(2))

        return GdlPool.getDistinct(arg1, arg2)

    def GdlFunction createFunction(SymbolList list)
	
        GdlConstant name = createConstant((SymbolAtom) list.get(0))

        List<GdlTerm> body = new ArrayList<GdlTerm>()
        for (int i = 1; i < list.size(); i++)
		
            body.add(createTerm(list.get(i)))

        return GdlPool.getFunction(name, body)

    def Gdl createGdl(Symbol symbol)
	
        if (symbol instanceof SymbolList)
		
            SymbolList list = (SymbolList) symbol
            SymbolAtom type = (SymbolAtom) list.get(0)

            if (type.getValue().equals("<="))
			
                return createRule(list)

        return createSentence(symbol)

    def GdlLiteral createLiteral(Symbol symbol)
	
        if (symbol instanceof SymbolList)
		
            SymbolList list = (SymbolList) symbol
            SymbolAtom type = (SymbolAtom) list.get(0)

            if (type.getValue().toLowerCase().equals("distinct"))
			
                return createDistinct(list)
            else if (type.getValue().toLowerCase().equals("not"))
			
                return createNot(list)
            else if (type.getValue().toLowerCase().equals("or"))
			
                return createOr(list)

        return createSentence(symbol)

    def GdlNot createNot(SymbolList list)
	
        return GdlPool.getNot(createLiteral(list.get(1)))

    def GdlOr createOr(SymbolList list)
	
        List<GdlLiteral> disjuncts = new ArrayList<GdlLiteral>()
        for (int i = 1; i < list.size(); i++)
		
            disjuncts.add(createLiteral(list.get(i)))

        return GdlPool.getOr(disjuncts)

    def GdlProposition createProposition(SymbolAtom atom)
	
        return GdlPool.getProposition(createConstant(atom))

    def GdlRelation createRelation(SymbolList list)
	
        GdlConstant name = createConstant((SymbolAtom) list.get(0))

        List<GdlTerm> body = new ArrayList<GdlTerm>()
        for (int i = 1; i < list.size(); i++)
		
            body.add(createTerm(list.get(i)))

        return GdlPool.getRelation(name, body)

    def GdlRule createRule(SymbolList list)
	
        GdlSentence head = createSentence(list.get(1))

        List<GdlLiteral> body = new ArrayList<GdlLiteral>()
        for (int i = 2; i < list.size(); i++)
		
            body.add(createLiteral(list.get(i)))

        return GdlPool.getRule(head, body)

    def GdlSentence createSentence(Symbol symbol)
	
        if (symbol instanceof SymbolAtom)
		
            return createProposition((SymbolAtom) symbol)
        else
		
            return createRelation((SymbolList) symbol)

    def static GdlTerm createTerm(String string) throws SymbolFormatException
	
        return createTerm(SymbolFactory.create(string))

    def static GdlTerm createTerm(Symbol symbol)
	
        if (symbol instanceof SymbolAtom)
		
            SymbolAtom atom = (SymbolAtom) symbol
            if (atom.getValue().charAt(0) == '?')
			
                return createVariable(atom)
            else
			
                return createConstant(atom)
        else
		
            return createFunction((SymbolList) symbol)

    def GdlVariable createVariable(SymbolAtom atom)
	
        return GdlPool.getVariable(atom.getValue())

File: ggpy/cruft/autocode/GdlFactory.py
#!/usr/bin/env python
""" generated source for module GdlFactory """
# package: org.ggp.base.util.gdl.factory
from symbol.grammar import Symbol, SymbolList, SymbolAtom

class GdlFactory(object):
    """ generated source for class GdlFactory """
    @classmethod
    def create(cls, string):
        """ generated source for method create """
        return cls.create(SymbolFactory.create(string))

    @classmethod
    def create_0(cls, symbol):
        """ generated source for method create_0 """
        try:
            return createGdl(symbol)
        except Exception as e:
            createGdl(symbol)
            raise GdlFormatException(symbol)

    @classmethod
    def createConstant(cls, atom):
        """ generated source for method createConstant """
        return GdlPool.getConstant(atom.getValue())

    @classmethod
    def createDistinct(cls, list_):
        """ generated source for method createDistinct """
        arg1 = createTerm(list_.get(1))
        arg2 = createTerm(list_.get(2))
        return GdlPool.getDistinct(arg1, arg2)

    @classmethod
    def createFunction(cls, list_):
        """ generated source for method createFunction """
        name = cls.createConstant(list_.get(0))
        body = ArrayList()
        i = 1
        while i < len(list_):
            body.add(createTerm(list_.get(i)))
            i += 1
        return GdlPool.getFunction(name, body)

    @classmethod
    def createGdl(cls, symbol):
        """ generated source for method createGdl """
        if isinstance(symbol, (SymbolList, )):
            if type_.getValue() == "<=":
                return createRule(list_)
        return createSentence(symbol)

    @classmethod
    def createLiteral(cls, symbol):
        """ generated source for method createLiteral """
        if isinstance(symbol, (SymbolList, )):
            if type_.getValue().lower() == "distinct":
                return cls.createDistinct(list_)
            elif type_.getValue().lower() == "not":
                return createNot(list_)
            elif type_.getValue().lower() == "or":
                return createOr(list_)
        return createSentence(symbol)

    @classmethod
    def createNot(cls, list_):
        """ generated source for method createNot """
        return GdlPool.getNot(cls.createLiteral(list_.get(1)))

    @classmethod
    def createOr(cls, list_):
        """ generated source for method createOr """
        disjuncts = ArrayList()
        i = 1
        while i < len(list_):
            disjuncts.add(cls.createLiteral(list_.get(i)))
            i += 1
        return GdlPool.getOr(disjuncts)

    @classmethod
    def createProposition(cls, atom):
        """ generated source for method createProposition """
        return GdlPool.getProposition(cls.createConstant(atom))

    @classmethod
    def createRelation(cls, list_):
        """ generated source for method createRelation """
        name = cls.createConstant(list_.get(0))
        body = ArrayList()
        i = 1
        while i < len(list_):
            body.add(createTerm(list_.get(i)))
            i += 1
        return GdlPool.getRelation(name, body)

    @classmethod
    def createRule(cls, list_):
        """ generated source for method createRule """
        head = createSentence(list_.get(1))
        body = ArrayList()
        i = 2
        while i < len(list_):
            body.add(cls.createLiteral(list_.get(i)))
            i += 1
        return GdlPool.getRule(head, body)

    @classmethod
    def createSentence(cls, symbol):
        """ generated source for method createSentence """
        if isinstance(symbol, (SymbolAtom, )):
            return cls.createProposition(symbol)
        else:
            return cls.createRelation(symbol)

    @classmethod
    @overloaded
    def createTerm(cls, string):
        """ generated source for method createTerm """
        return cls.createTerm(SymbolFactory.create(string))

    @classmethod
    @createTerm.register(object, Symbol)
    def createTerm_0(cls, symbol):
        """ generated source for method createTerm_0 """
        if isinstance(symbol, (SymbolAtom, )):
            if atom.getValue().charAt(0) == '?':
                return createVariable(atom)
            else:
                return cls.createConstant(atom)
        else:
            return cls.createFunction(symbol)

    @classmethod
    def createVariable(cls, atom):
        """ generated source for method createVariable """
        return GdlPool.getVariable(atom.getValue())

